From Copilot:

## PR #2950 Analysis: Remaining Leaks and New Issues

### ✅ What PR #2950 Successfully Fixes

Comparing against the issues identified in the prior review of PR #2668:

| Prior Issue | Status |
|---|---|
| `send_cleanup_signal()` crashes on `None` | ✅ Fixed — null guards added (L941-942) |
| Unbounded atexit handler accumulation | ✅ Fixed — `_atexit_registered` flag prevents re-registration (L886-889) |
| `del self.recovery_queue` without `close()` in `prepare_process()` | ✅ Fixed — `close()` + `join_thread()` called before `del` (L879-882) |
| `err.message` AttributeError in `remove_workers()` | ✅ Fixed — rewrote to catch `AttributeError` and use `str(e)` (L254-255) |
| `remove_workers()` not protected by lock | ✅ Fixed — `with cls.lock:` added (L248) |
| `ctx` may be `None` in `clean_log_recovery_process()` | ✅ Fixed — guard added (L815-816) |
| `join()` without timeout at experiment end (autosubmit.py) | ✅ Fixed — `timeout=60` added (L2379) |
| `send_cleanup_signal()` doesn't call `terminate()` after join timeout | ✅ Fixed — added SIGTERM fallback (L947-950) |

---

### 🔴 Critical Remaining Issues

#### Issue 1: Double-start race condition in `create_new_process()` — NEW BUG (L892-904)

```python
def create_new_process(self, new_platform, as_conf) -> None:
    try:
        self.log_recovery_process = self.ctx.Process(...)
        if self.log_recovery_process is not None:  # Dead code: ctx.Process() never returns None
            self.log_recovery_process.start()
    except Exception as e:
        Log.error(f"Failed to start log recovery process: {e}")
        self.log_recovery_process = None
    if self.log_recovery_process is not None and not self.log_recovery_process.is_alive():
        self.log_recovery_process.start()  # ← BUG: tries to start an already-started process!
```

The final two lines constitute a regression bug introduced by this PR. After `start()` succeeds inside the `try` block, `is_alive()` is checked immediately. If the spawned process crashes very quickly during startup (e.g., due to misconfiguration, missing resources), `is_alive()` returns `False`, and the code tries to call `.start()` a **second time** on an already-started process. This raises `AssertionError: cannot start a process twice` in CPython's multiprocessing module.

Additionally, `ctx.Process()` **never** returns `None` — it either returns a Process object or raises an exception. The `if self.log_recovery_process is not None:` check inside the try block is dead code.

**Fix**: Remove the final `if not is_alive()` block entirely. The re-start logic is both incorrect and dangerous:
```python
def create_new_process(self, new_platform, as_conf) -> None:
    try:
        self.log_recovery_process = self.ctx.Process(
            target=recover_platform_job_logs_wrapper,
            args=(new_platform, self.recovery_queue, self.work_event, self.cleanup_event, as_conf),
            name=f"{self.name}_log_recovery")
        self.log_recovery_process.daemon = True
        self.log_recovery_process.start()
    except Exception as e:
        Log.error(f"Failed to start log recovery process: {e}")
        self.log_recovery_process = None
```

---

#### Issue 2: Process not killed at experiment end when `join()` times out (autosubmit.py L2376-2381)

```python
for p in platforms_to_test:
    if p.log_recovery_process and p.log_recovery_process.is_alive():
        p.cleanup_event.set()
        p.log_recovery_process.join(timeout=60)
        if p.log_recovery_process.is_alive():
            Log.warning(f"Process for platform {p.name} failed to terminate within the timeout.")
            # ← Missing: p.log_recovery_process.terminate()
```

When the process doesn't stop within 60 seconds, only a warning is logged but the process is left running indefinitely. This is an orphan process leak. `clean_log_recovery_process()` has the correct pattern (warn + `terminate()` + second join), but this exit path in `autosubmit.py` does not.

---

### 🟡 Moderate Remaining Issues

#### Issue 3: `clean_log_recovery_process()` leaves inconsistent state (L809-841)

After `clean_log_recovery_process()` returns, the Platform object is in a partially-cleaned state:
- `self.cleanup_event` → still the old, **SET** event (not reset)
- `self.log_recovery_process` → still references the dead process object (not set to None)
- `self.recovery_queue` → closed queue (not replaced)
- `self.work_event` → newly created Event (intermediate, will be overwritten by `prepare_process()`)

The normal call sequence (`clean_log_recovery_process()` → `prepare_process()`) handles this because `prepare_process()` overwrites `cleanup_event`, creates a new queue, etc. But the **intermediate state is inconsistent**, meaning:
- If code checks `p.log_recovery_process` between the two calls, it finds a dead process and may make wrong decisions
- If `cleanup_event.is_set()` is checked in this window, it returns `True` (stale), which could cause the new child process to exit immediately
- If process creation fails in `create_new_process()`, `log_recovery_process` is set to `None`, but `cleanup_event` and `recovery_queue` are left in the post-`clean_log_recovery_process()` inconsistent state

**Fix**: Reset `cleanup_event`, `log_recovery_process`, and `recovery_queue` at the end of `clean_log_recovery_process()`:
```python
self.cleanup_event = None
self.log_recovery_process = None
self.recovery_queue = None
```

#### Issue 4: Intermediate `work_event` Event/semaphore wastefully created (L840)

`clean_log_recovery_process()` ends by creating a new `work_event`:
```python
self.work_event = self.ctx.Event()  # ← Creates semaphore
```

Then `prepare_process()` immediately overwrites it with another new one:
```python
self.work_event = self.ctx.Event()  # ← Creates another semaphore
```

The intermediate Event from `clean_log_recovery_process()` is discarded without being explicitly closed. Its underlying OS semaphore is not freed until Python's resource tracker or GC runs — this is exactly the class of semaphore leak the PR is trying to fix. The fix is to set `self.work_event = None` in `clean_log_recovery_process()` (after removing it from `worker_events`) and let `prepare_process()` create the only Event.

#### Issue 5: `update_workers()` not protected by `cls.lock` (L238-241)

`remove_workers()` now correctly uses `with cls.lock:` but `update_workers()` still mutates the shared class-level list without any lock:
```python
@classmethod
def update_workers(cls, event_worker):
    cls.worker_events.append(event_worker)  # ← No lock!
```

If `update_workers()` is called concurrently (e.g., during a platform spawn) while `remove_workers()` is iterating the list, this is an unsynchronized mutation. The lock is already defined and used for removal — it should also cover appends.

#### Issue 6: Missing SIGKILL fallback when SIGTERM fails in `clean_log_recovery_process()` (L828-829)

```python
if self.log_recovery_process.is_alive():
    Log.error("Log recovery process refused to terminate")
    # ← Missing: self.log_recovery_process.kill()
```

When a process ignores SIGTERM and `join(timeout=60)` expires a second time, only an error is logged but the process remains alive forever. `kill()` (SIGKILL) should be called as a last resort to ensure the process is actually terminated.

---

### 🟢 Minor Issues

#### Issue 7: `wait_mandatory_time()` catches wrong exception (L967)

```python
except queue.Empty:
    Log.warning(f"Queue is empty and process log is {process_log} seconds.")
```

`queue.Empty` is raised by `Queue.get()` / `Queue.get_nowait()`, **not** by `Queue.empty()`. The `multiprocessing.Queue.empty()` call could raise `OSError` or other system-level errors under stress. This `except queue.Empty` clause is dead code that will never fire and masks the real potential error. The exception type should be `OSError` or simply `Exception`.

Also, the warning message itself is incorrect — `process_log` is a `bool`, so `f"{process_log} seconds"` prints `"True seconds"` or `"False seconds"`.

#### Issue 8: `_atexit_registered = None` instead of `False` in `__init__` (L151)

```python
self._atexit_registered = None
```

The subsequent check `if not getattr(self, "_atexit_registered", False)` works because `not None` is `True`, but semantically this should be `False` from the start. This also means the `getattr(..., False)` default will never be hit (since `_atexit_registered` is always set in `__init__`), so the `getattr` pattern could simply be `if not self._atexit_registered`.

#### Issue 9: `daemon = True` missing from new `create_new_process()` (L894)

The PR's new version of `create_new_process()` does not set `self.log_recovery_process.daemon = True`. The original code in PR #2668 had `self.log_recovery_process.daemon = True` before `start()`. Without this, if the parent process is killed while the child is running, the child may continue as an orphan. (This is a regression from PR #2668's implementation.)

---

### Summary Table

| # | Location | Issue | Severity | Type |
|---|---|---|---|---|
| 1 | `platform.py:903-904` | Double-start `AssertionError` bug introduced by this PR | 🔴 Critical | Regression / crash |
| 2 | `autosubmit.py:2380-2381` | Process not killed after join timeout at experiment end | 🔴 Critical | Orphan process leak |
| 3 | `platform.py:809-841` | `clean_log_recovery_process()` leaves inconsistent state (stale `cleanup_event`, `log_recovery_process`, `recovery_queue`) | 🟡 Moderate | State inconsistency |
| 4 | `platform.py:840` | Intermediate `work_event` semaphore created then discarded | 🟡 Moderate | Semaphore leak |
| 5 | `platform.py:238-241` | `update_workers()` not under `cls.lock` | 🟡 Moderate | Race condition |
| 6 | `platform.py:828-829` | No SIGKILL when SIGTERM fails in `clean_log_recovery_process()` | 🟡 Moderate | Zombie process |
| 7 | `platform.py:967` | Wrong exception caught in `wait_mandatory_time()` | 🟢 Minor | Dead code / misleading |
| 8 | `platform.py:151` | `_atexit_registered = None` should be `False` | 🟢 Minor | Semantics |
| 9 | `platform.py:894` | `daemon = True` not set on new process | 🟢 Minor | Regression |

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -43,6 +43,7 @@
 - Slurm tests in CICD are using a new container by @giovtorres #2871
 - Fixed main-loop iteration message handling singular/plural and reducing number of line breaks #2969
 - Updated `Defining the workflow` documentation section #2954
+- Improving the handling of leakages that were happening when dealing with queues and processes #2900
 
 ### 4.1.16: Postgres (experimental) support, bug fixes, and enhancements
 

--- a/autosubmit/autosubmit.py
+++ b/autosubmit/autosubmit.py
@@ -2374,10 +2374,16 @@ class Autosubmit:
                     Autosubmit.restore_platforms(platforms_to_test, as_conf=as_conf, expid=expid)
                 Log.info("Waiting for all logs to be updated")
                 for p in platforms_to_test:
-                    if p.log_recovery_process and p.log_recovery_process.is_alive():
+                    if (p.log_recovery_process is not None and p.log_recovery_process.is_alive()
+                            and p.cleanup_event is not None):
                         p.cleanup_event.set()  # Send cleanup event
                         p.log_recovery_process.join(timeout=60)
                         if p.log_recovery_process.is_alive():
+                            p.log_recovery_process.terminate()
+                            p.log_recovery_process.join(timeout=60)
+                            if p.log_recovery_process.is_alive():
+                                Log.error("Log recovery process refused to terminate")
+                                p.log_recovery_process.kill()
                             Log.warning(f"Process for platform {p.name} failed to terminate within the timeout.")
                 Autosubmit.check_logs_status(job_list, as_conf, new_run=False)
                 job_list.save()
@@ -2986,13 +2992,13 @@ class Autosubmit:
         submitter.load_platforms(as_conf)
         platforms = submitter.platforms
 
-        platforms_to_test = set()
+        platforms_to_test: list['ParamikoPlatform'] = list()
         for job in job_list.get_job_list():
             job.submitter = submitter
             if not job.platform_name:
                 job.platform_name = hpcarch
             job.platform = platforms[job.platform_name]
-            platforms_to_test.add(job.platform)
+            platforms_to_test.append(job.platform)
 
         completed_jobnames = Autosubmit.online_recovery(as_conf, platforms_to_test, job_list, offline)
         current_active_jobs = job_list.get_in_queue()
@@ -3314,7 +3320,7 @@ class Autosubmit:
 
                 as_conf = AutosubmitConfig(experiment_id, BasicConfig, YAMLParserFactory())
                 as_conf.check_conf_files(False, no_log=True)
-                user = os.stat(as_conf.conf_folder_yaml).st_uid
+                user = int(Path(as_conf.conf_folder_yaml).stat().st_uid)
                 try:
                     user = pwd.getpwuid(user).pw_name
                 except Exception:
@@ -3322,7 +3328,7 @@ class Autosubmit:
                         "The user does not exist anymore in the system, using id instead")
                     continue
 
-                created = datetime.datetime.fromtimestamp(os.path.getmtime(as_conf.conf_folder_yaml))
+                created = datetime.datetime.fromtimestamp(Path(as_conf.conf_folder_yaml).stat().st_mtime)
 
                 if as_conf.get_svn_project_url():
                     model = as_conf.get_svn_project_url()
@@ -3502,8 +3508,10 @@ class Autosubmit:
 
                     sep = '\n\t'
                     Log.result(sep.join(['Directories configured successfully:'] + [str(d) for d in dirs]))
-                except (IOError, OSError) as e:
+                except (IOError) as e:
                     raise AutosubmitCritical(f"Can not write config file: {e.message}", 7012)
+                except (OSError) as e:
+                    raise AutosubmitCritical(f"Can not write config file: {e}", 7012)
         except (AutosubmitCritical, AutosubmitError):
             raise
         except BaseException as e:
@@ -3606,7 +3614,7 @@ class Autosubmit:
             # Look for %_%
             variables = re.findall('%(?<!%%)[a-zA-Z0-9_.-]+%(?!%%)', template_content, flags=re.IGNORECASE)
             variables = [variable[1:-1].upper() for variable in variables]
-            results = {}
+            results: dict = dict()
             # Change format
             for old_format_key in variables:
                 for key in as_conf.load_parameters().keys():
@@ -3996,7 +4004,7 @@ class Autosubmit:
         :return: ``True`` if the experiment has been successfully archived. ``False`` otherwise.
         :rtype: bool
         """
-        exp_folder = os.path.join(BasicConfig.LOCAL_ROOT_DIR, expid)
+        exp_folder = Path(BasicConfig.LOCAL_ROOT_DIR).joinpath(expid)
 
         if not noclean:
             # Cleaning to reduce file size.
@@ -4007,29 +4015,27 @@ class Autosubmit:
 
         # Getting year of last completed. If not, year of expid folder
         year = None
-        tmp_folder = os.path.join(exp_folder, BasicConfig.LOCAL_TMP_DIR)
-        if os.path.isdir(tmp_folder):
-            for filename in os.listdir(tmp_folder):
-                if filename.endswith("COMPLETED"):
-                    file_year = time.localtime(os.path.getmtime(
-                        os.path.join(tmp_folder, filename))).tm_year
+        tmp_folder = exp_folder.joinpath(BasicConfig.LOCAL_TMP_DIR)
+        if tmp_folder.is_dir():
+            for filename in tmp_folder.iterdir():
+                if filename.name.endswith("COMPLETED"):
+                    file_year = time.localtime(tmp_folder.joinpath(filename).stat().st_mtime).tm_year
                     if year is None or year < file_year:
                         year = file_year
 
         if year is None:
             year = time.localtime(os.path.getmtime(exp_folder)).tm_year
         try:
-            year_path = os.path.join(BasicConfig.LOCAL_ROOT_DIR, str(year))
-            if not os.path.exists(year_path):
-                os.mkdir(year_path)
-                os.chmod(year_path, 0o775)
+            year_path = Path(BasicConfig.LOCAL_ROOT_DIR).joinpath(str(year))
+            if not year_path.exists():
+                year_path.mkdir(mode=0o775, parents=True)
         except Exception as e:
             raise AutosubmitCritical(f"Failed to create year-directory {str(year)} for experiment {expid}", 7012,
                                      str(e))
         Log.info(f"Archiving in year {str(year)}")
 
         if rocrate:
-            Autosubmit.rocrate(expid, Path(year_path))
+            Autosubmit.rocrate(expid, year_path)
             Log.info('RO-Crate ZIP file created!')
         else:
             # Creating tar file
@@ -4041,10 +4047,11 @@ class Autosubmit:
                 else:
                     compress_type = "w"
                     output_filepath = f'{expid}.tar'
-                with tarfile.open(os.path.join(year_path, output_filepath), compress_type) as tar:
+                year_path_file = year_path.joinpath(output_filepath)
+                with tarfile.open(year_path_file, compress_type) as tar:
                     tar.add(exp_folder, arcname='')
                     tar.close()
-                    os.chmod(os.path.join(year_path, output_filepath), 0o775)
+                    year_path_file.chmod(mode=0o775)
             except Exception as e:
                 raise AutosubmitCritical("Can not write tar file", 7012, str(e))
 
@@ -4080,11 +4087,11 @@ class Autosubmit:
         :param rocrate: flag to enable RO-Crate
         :type rocrate: bool
         """
-        exp_folder = os.path.join(BasicConfig.LOCAL_ROOT_DIR, experiment_id)
+        exp_folder = Path(BasicConfig.LOCAL_ROOT_DIR, experiment_id)
 
         # Searching by year. We will store it on database
         year = datetime.datetime.today().year
-        archive_path = None
+        archive_path = Path()
         if rocrate:
             compress_type = None
             output_pathfile = f'{experiment_id}.zip'
@@ -4095,9 +4102,8 @@ class Autosubmit:
             compress_type = "r:"
             output_pathfile = f'{experiment_id}.tar'
         while year > 2000:
-            archive_path = os.path.join(
-                BasicConfig.LOCAL_ROOT_DIR, str(year), output_pathfile)
-            if os.path.exists(archive_path):
+            archive_path = Path(BasicConfig.LOCAL_ROOT_DIR).joinpath(str(year), output_pathfile)
+            if archive_path.exists():
                 break
             year -= 1
 
@@ -4108,15 +4114,14 @@ class Autosubmit:
 
         # Creating tar file
         Log.info("Unpacking tar file ... ")
-        if not os.path.isdir(exp_folder):
-            os.mkdir(exp_folder)
+        exp_folder.mkdir(exist_ok=True, parents=True)
         try:
             if rocrate:
                 import zipfile
-                with zipfile.ZipFile(archive_path, 'r') as zip:
+                with zipfile.ZipFile(str(archive_path), 'r') as zip:
                     zip.extractall(exp_folder)
             else:
-                with tarfile.open(os.path.join(archive_path), compress_type) as tar:
+                with tarfile.open(archive_path, compress_type) as tar:
                     tar.extractall(exp_folder)
                     tar.close()
         except Exception as e:
@@ -4127,7 +4132,7 @@ class Autosubmit:
         Log.info("Unpacking finished")
 
         try:
-            os.remove(archive_path)
+            archive_path.unlink()
         except Exception as e:
             Log.printlog(f"Can not remove archived file folder: {str(e)}", 7012)
             Log.result(f"Experiment {experiment_id} unarchived successfully")
@@ -5735,35 +5740,35 @@ class Autosubmit:
         truthy_values = ["true", "yes", "y", "1", ""]
         if not force_all:
             expid_list: list[str] = [
-                expid
-                for expid in expid_list
+                expid_in_list
+                for expid_in_list in expid_list
                 if force_yes or input(
-                    f"Confirm stopping the experiment: {expid} (y/n)[enter=y]? ").lower() in truthy_values
+                    f"Confirm stopping the experiment: {expid_in_list} (y/n)[enter=y]? ").lower() in truthy_values
             ]
 
         sig_to_process = signal.SIGKILL if force else signal.SIGINT
         killed_expids = []
-        for expid in expid_list:
-            pid: int = process_id(expid)
+        for expid_in_list in expid_list:
+            pid: int = process_id(expid_in_list)
             if not pid or pid <= 1:
-                Log.info(f"Expid {expid} was not running")
+                Log.info(f"Expid {expid_in_list} was not running")
                 continue
             try:
                 os.kill(pid, sig_to_process)
-                killed_expids.append(expid)
+                killed_expids.append(expid_in_list)
             except Exception as e:
-                Log.warning(f"An error occurred while stopping the autosubmit process for expid '{expid}': {str(e)}")
+                Log.warning(f"An error occurred while stopping the autosubmit process for expid '{expid_in_list}': {str(e)}")
 
-        for expid in killed_expids:
+        for expid_in_list in killed_expids:
             if not force:
-                Log.info(f"Checking the status of the expid: {expid}")
+                Log.info(f"Checking the status of the expid: {expid_in_list}")
                 while True:
-                    if not process_id(expid):
-                        Log.info(f"Expid {expid} is stopped")
+                    if not process_id(expid_in_list):
+                        Log.info(f"Expid {expid_in_list} is stopped")
                         break
-                    Log.info(f"Waiting for the autosubmit run to safety stop: {expid}")
+                    Log.info(f"Waiting for the autosubmit run to safety stop: {expid_in_list}")
                     sleep(5)
             if cancel:
-                job_list, _, _, _, _, _, _, _ = Autosubmit.prepare_run(expid, check_scripts=False)
+                job_list, _, _, _, _, _, _, _ = Autosubmit.prepare_run(expid_in_list, check_scripts=False)
                 cancel_jobs(job_list, active_jobs_filter=current_status, target_status=status)
         return True

--- a/autosubmit/autosubmit.py
+++ b/autosubmit/autosubmit.py
@@ -2368,6 +2368,7 @@ class Autosubmit:
                     Log.info("Connecting to the platforms, to recover missing logs")
                     submitter = ParamikoSubmitter(as_conf=as_conf)
                     if submitter.platforms is None:
+                        Log.error("Failed to retrieve platforms configuration from ParamikoSubmitter.")
                         raise AutosubmitCritical("No platforms configured!!!", 7014)
                     platforms_to_test = [value for value in submitter.platforms.values()]
                     Autosubmit.restore_platforms(platforms_to_test, as_conf=as_conf, expid=expid)
@@ -2375,14 +2376,18 @@ class Autosubmit:
                 for p in platforms_to_test:
                     if p.log_recovery_process and p.log_recovery_process.is_alive():
                         p.cleanup_event.set()  # Send cleanup event
-                        p.log_recovery_process.join()
+                        p.log_recovery_process.join(timeout=60)
+                        if p.log_recovery_process.is_alive():
+                            Log.warning(f"Process for platform {p.name} failed to terminate within the timeout.")
                 Autosubmit.check_logs_status(job_list, as_conf, new_run=False)
                 job_list.save()
-                if len(job_list.get_completed_failed_without_logs()) == 0:
+                failed_jobs = job_list.get_completed_failed_without_logs()
+                if len(failed_jobs) == 0:
                     Log.result("Autosubmit recovered all job logs.")
                 else:
                     Log.warning(
-                        f"Autosubmit couldn't recover the following job logs: {[job.name for job in job_list.get_completed_failed_without_logs()]}")
+                        f"Autosubmit couldn't recover the following job logs: {[job.name for job in failed_jobs]}")
+                    Log.warning(f"Failed jobs details: {', '.join([str(job) for job in failed_jobs])}")
                 try:
                     exp_history = ExperimentHistory(expid)
                     exp_history.process_job_list_changes_to_experiment_totals(job_list.get_job_list())

--- a/autosubmit/platforms/platform.py
+++ b/autosubmit/platforms/platform.py
@@ -148,6 +148,8 @@ class Platform:
         :param auth_password: Optional password for two-factor authentication.
         :type auth_password: str or list, optional
         """
+        self._atexit_registered = None
+        self.processed_wrapper_logs = None
         self.ctx = self.get_mp_context()
         self.connected = False
         self.expid: str = expid
@@ -241,12 +243,16 @@ class Platform:
     @classmethod
     def remove_workers(cls, event_worker: Event) -> None:
         """Remove the given even worker from the list of workers in this class. """
-        if event_worker in cls.worker_events:
+        if event_worker is None:
+            return
+        with cls.lock:
             try:
-                event_worker.set()
-            except Exception as err:
-                Log.warning(f"Calling the worker raised an issue: {err.message}")
-            cls.worker_events.remove(event_worker)
+                if event_worker in cls.worker_events:
+                    cls.worker_events.remove(event_worker)
+            except ValueError:
+                pass
+            except AttributeError as e:
+                Log.warning(f"The event couldn't be removed, event has an invalid state: \n{str(e)}")
 
     @property
     @autosubmit_parameter(name='current_arch')
@@ -806,26 +812,32 @@ class Platform:
         This method sets the cleanup event to signal the log recovery process to finish,
         waits for the process to join with a timeout, and then resets all related variables.
         """
-        if self.cleanup_event:
-            self.cleanup_event.set()  # Indicates to old child ( if reachable ) to finish.
+        if self.ctx is None:
+            self.ctx = self.get_mp_context()
 
-        if self.log_recovery_process and self.log_recovery_process.is_alive():
-            self.log_recovery_process.terminate()
+        if self.cleanup_event:
+            self.cleanup_event.set()
+
+        if self.log_recovery_process:
             self.log_recovery_process.join(timeout=60)
 
-        # Resets everything related to the log recovery process.
-        if self.recovery_queue is not None:
-            self.recovery_queue.close()
-            self.recovery_queue.join_thread()
+            if self.log_recovery_process.is_alive():
+                self.log_recovery_process.terminate()
+                self.log_recovery_process.join(timeout=60)
 
-        self.recovery_queue = self.ctx.Queue()
+                if self.log_recovery_process.is_alive():
+                    Log.error("Log recovery process refused to terminate")
+
+        if self.recovery_queue:
+            try:
+                self.recovery_queue.close()
+                self.recovery_queue.join_thread()
+            except Exception as e:
+                Log.warning(f"Queue cleanup failed: {e}")
         self.log_retrieval_process_active = False
-        if self.work_event:
+        if self.work_event is not None:
             self.remove_workers(self.work_event)
         self.work_event = self.ctx.Event()
-        self.cleanup_event = self.ctx.Event()
-        self.log_recovery_process = None
-        self.log_retrieval_process_active = False
         self.processed_wrapper_logs = set()
 
     def load_process_info(self, platform):
@@ -858,35 +870,53 @@ class Platform:
 
     def prepare_process(self) -> 'Platform':
         new_platform = self.create_a_new_copy()
+        if self.ctx is None:
+            self.ctx = self.get_mp_context()
         self.work_event = self.ctx.Event()
         self.cleanup_event = self.ctx.Event()
         Platform.update_workers(self.work_event)
         self.load_process_info(new_platform)
-        if self.recovery_queue:
+        if self.recovery_queue is not None:
+            self.recovery_queue.close()
+            self.recovery_queue.join_thread()
             del self.recovery_queue
         # Retrieval log process variables
         self.recovery_queue = CopyQueue(ctx=self.ctx)
         # Cleanup will be automatically prompt on control + c or a normal exit
-        atexit.register(self.send_cleanup_signal)
-        atexit.register(self.close_connection)
+        if not getattr(self, "_atexit_registered", False):
+            atexit.register(self.send_cleanup_signal)
+            atexit.register(self.close_connection)
+            self._atexit_registered = True
         return new_platform
 
     def create_new_process(self, new_platform: 'Platform', as_conf) -> None:
-        self.log_recovery_process = self.ctx.Process(
-            target=recover_platform_job_logs_wrapper,
-            args=(new_platform, self.recovery_queue, self.work_event, self.cleanup_event, as_conf),
-            name=f"{self.name}_log_recovery")
-        self.log_recovery_process.daemon = True
-        self.log_recovery_process.start()
+        try:
+            self.log_recovery_process = self.ctx.Process(
+                target=recover_platform_job_logs_wrapper,
+                args=(new_platform, self.recovery_queue, self.work_event, self.cleanup_event, as_conf),
+                name=f"{self.name}_log_recovery")
+            if self.log_recovery_process is not None:
+                self.log_recovery_process.start()
+        except Exception as e:
+            Log.error(f"Failed to start log recovery process: {e}")
+            self.log_recovery_process = None
+        if self.log_recovery_process is not None and not self.log_recovery_process.is_alive():
+            self.log_recovery_process.start()
 
-    @staticmethod
-    def get_mp_context():
-        return multiprocessing.get_context('spawn')
+    def get_mp_context(self):
+        if not hasattr(self, 'ctx') or self.ctx is None:
+            self.ctx = multiprocessing.get_context('spawn')
+        return self.ctx
 
     def join_new_process(self):
-        # Prevents zombies
-        os.waitpid(self.log_recovery_process.pid, os.WNOHANG)
-        Log.result(f"Process {self.log_recovery_process.name} started with pid {self.log_recovery_process.pid}")
+        # Check if the process is finished; prevent zombies
+        if self.log_recovery_process is not None:
+            ret_pid, ret_status = os.waitpid(self.log_recovery_process.pid, os.WNOHANG)
+            if ret_pid == 0:  # Process is still running
+                Log.info(f"Process {self.log_recovery_process.pid} is still running.")
+                # Optionally, handle cleanup here (terminate / join)
+            else:
+                Log.result(f"Process {self.log_recovery_process.name} finished with pid {self.log_recovery_process.pid}")
 
     def spawn_log_retrieval_process(self, as_conf: Optional['AutosubmitConfig']) -> None:
         """Spawns a process to recover the logs of the jobs that have been completed on this platform.
@@ -908,10 +938,16 @@ class Platform:
         """Sends a cleanup signal to the log recovery process if it is alive.
         This function is executed by the atexit module
         """
-        if self.log_recovery_process.is_alive():
+        if (self.work_event is not None and self.cleanup_event is not None and
+                self.log_recovery_process is not None and self.log_recovery_process.is_alive()):
             self.work_event.clear()
             self.cleanup_event.set()
             self.log_recovery_process.join(timeout=60)
+
+            if self.log_recovery_process.is_alive():
+                Log.warning(f"Process {self.log_recovery_process.pid} didn't terminate within the timeout.")
+                self.log_recovery_process.terminate()
+                self.log_recovery_process.join(timeout=60)
 
     def wait_mandatory_time(self, sleep_time: int = 60) -> bool:
         """Waits for the work_event to be set or the cleanup_event to be set for a mandatory time.
@@ -922,18 +958,20 @@ class Platform:
         :rtype: bool
         """
         process_log = False
-        for remaining in range(sleep_time, 0, -1):
-            time.sleep(1)
-            if self.work_event.is_set() or not self.recovery_queue.empty():
-                process_log = True
-            if self.cleanup_event.is_set():
-                process_log = True
-                break
+        for _ in range(0, sleep_time, 5):
+            time.sleep(5)
+            try:
+                if self.work_event.is_set() or not self.recovery_queue.empty() or self.cleanup_event.is_set():
+                    process_log = True
+                    break
+            except queue.Empty:
+                Log.warning(f"Queue is empty and process log is {process_log} seconds.")
+        if not process_log:
+            Log.warning(f"No work found within {sleep_time} seconds.")
         return process_log
 
     def wait_for_work(self, sleep_time: int = 60) -> bool:
-        """
-        Waits a mandatory time and then waits until there is work, no work to more process or the cleanup event is set.
+        """Waits a mandatory time and then waits until there is work, no work to more process or the cleanup event is set.
 
         :param sleep_time: Maximum time to wait in seconds. Defaults to 60.
         :type sleep_time: int
@@ -942,22 +980,8 @@ class Platform:
         """
         process_log = self.wait_mandatory_time(sleep_time)
         if not process_log:
-            process_log = self.wait_until_timeout(self.keep_alive_timeout - sleep_time)
+            process_log = self.wait_mandatory_time(self.keep_alive_timeout - sleep_time)
         self.work_event.clear()
-        return process_log
-
-    def wait_until_timeout(self, timeout: int = 60) -> bool:
-        """Waits until the timeout is reached or any signal is set to process logs.
-
-        :param timeout: Maximum time to wait in seconds. Defaults to 60.
-        :return: True if there is work to process, False otherwise.
-        """
-        process_log = False
-        for _ in range(timeout, 0, -1):
-            time.sleep(1)
-            if self.work_event.is_set() or not self.recovery_queue.empty() or self.cleanup_event.is_set():
-                process_log = True
-                break
         return process_log
 
     def recover_job_log(self, identifier: str, jobs_pending_to_process: set[Any],

--- a/autosubmit/platforms/platform.py
+++ b/autosubmit/platforms/platform.py
@@ -148,7 +148,7 @@ class Platform:
         :param auth_password: Optional password for two-factor authentication.
         :type auth_password: str or list, optional
         """
-        self._atexit_registered = None
+        self._atexit_registered = False
         self.processed_wrapper_logs = None
         self.ctx = self.get_mp_context()
         self.connected = False
@@ -237,8 +237,17 @@ class Platform:
 
     @classmethod
     def update_workers(cls, event_worker):
-        # This is visible on all instances simultaneosly. Is to send the keep alive signal.
-        cls.worker_events.append(event_worker)
+        # This is visible on all instances simultaneously. Is to send the keep alive signal.
+        if event_worker is None:
+            return
+        with cls.lock:
+            try:
+                if event_worker in cls.worker_events:
+                    cls.worker_events.append(event_worker)
+            except ValueError:
+                pass
+            except AttributeError as e:
+                Log.warning(f"The event couldn't be stored, event has an invalid state: \n{str(e)}")
 
     @classmethod
     def remove_workers(cls, event_worker: Event) -> None:
@@ -777,7 +786,6 @@ class Platform:
         """
         raise NotImplementedError  # pragma: no cover
 
-
     def add_job_to_log_recover(self, job):
         if job.id and int(job.id) != 0:
             self.recovery_queue.put(job)
@@ -824,9 +832,15 @@ class Platform:
             if self.log_recovery_process.is_alive():
                 self.log_recovery_process.terminate()
                 self.log_recovery_process.join(timeout=60)
-
                 if self.log_recovery_process.is_alive():
                     Log.error("Log recovery process refused to terminate")
+                    self.log_recovery_process.kill()
+
+            if hasattr(self.log_recovery_process, "close"):
+                try:
+                    self.log_recovery_process.close()
+                except Exception as e:
+                    Log.warning(f"Failed to close process: {e}")
 
         if self.recovery_queue:
             try:
@@ -834,10 +848,20 @@ class Platform:
                 self.recovery_queue.join_thread()
             except Exception as e:
                 Log.warning(f"Queue cleanup failed: {e}")
+
         self.log_retrieval_process_active = False
+
+        # Reuse existing Event objects instead of destroying them.
+        # Clearing the events is sufficient to give the next child process a clean initial state.
+        if self.cleanup_event is not None:
+            self.cleanup_event.clear()
         if self.work_event is not None:
-            self.remove_workers(self.work_event)
-        self.work_event = self.ctx.Event()
+            self.work_event.clear()
+
+        self.cleanup_event = None
+        self.recovery_queue = None
+        self.log_recovery_process = None
+        self.work_event = None
         self.processed_wrapper_logs = set()
 
     def load_process_info(self, platform):
@@ -872,9 +896,14 @@ class Platform:
         new_platform = self.create_a_new_copy()
         if self.ctx is None:
             self.ctx = self.get_mp_context()
-        self.work_event = self.ctx.Event()
-        self.cleanup_event = self.ctx.Event()
-        Platform.update_workers(self.work_event)
+
+        # Allocate Events only on the first call.
+        # On subsequent respawn cycles the events are reused (cleared by clean_log_recovery_process)
+        if self.work_event is None:
+            self.work_event = self.ctx.Event()
+            Platform.update_workers(self.work_event)
+        if self.cleanup_event is None:
+            self.cleanup_event = self.ctx.Event()
         self.load_process_info(new_platform)
         if self.recovery_queue is not None:
             self.recovery_queue.close()
@@ -895,13 +924,11 @@ class Platform:
                 target=recover_platform_job_logs_wrapper,
                 args=(new_platform, self.recovery_queue, self.work_event, self.cleanup_event, as_conf),
                 name=f"{self.name}_log_recovery")
-            if self.log_recovery_process is not None:
-                self.log_recovery_process.start()
+            self.log_recovery_process.daemon = True
+            self.log_recovery_process.start()
         except Exception as e:
             Log.error(f"Failed to start log recovery process: {e}")
             self.log_recovery_process = None
-        if self.log_recovery_process is not None and not self.log_recovery_process.is_alive():
-            self.log_recovery_process.start()
 
     def get_mp_context(self):
         if not hasattr(self, 'ctx') or self.ctx is None:
@@ -914,9 +941,10 @@ class Platform:
             ret_pid, ret_status = os.waitpid(self.log_recovery_process.pid, os.WNOHANG)
             if ret_pid == 0:  # Process is still running
                 Log.info(f"Process {self.log_recovery_process.pid} is still running.")
-                # Optionally, handle cleanup here (terminate / join)
             else:
                 Log.result(f"Process {self.log_recovery_process.name} finished with pid {self.log_recovery_process.pid}")
+        else:
+            Log.result("Log_Recovery_Process is empty no process joinned")
 
     def spawn_log_retrieval_process(self, as_conf: Optional['AutosubmitConfig']) -> None:
         """Spawns a process to recover the logs of the jobs that have been completed on this platform.
@@ -949,6 +977,11 @@ class Platform:
                 self.log_recovery_process.terminate()
                 self.log_recovery_process.join(timeout=60)
 
+        if self.work_event is not None:
+            self.remove_workers(self.work_event)
+            self.work_event = None
+            self.cleanup_event = None
+
     def wait_mandatory_time(self, sleep_time: int = 60) -> bool:
         """Waits for the work_event to be set or the cleanup_event to be set for a mandatory time.
 
@@ -964,8 +997,8 @@ class Platform:
                 if self.work_event.is_set() or not self.recovery_queue.empty() or self.cleanup_event.is_set():
                     process_log = True
                     break
-            except queue.Empty:
-                Log.warning(f"Queue is empty and process log is {process_log} seconds.")
+            except OSError:
+                Log.warning(f"Queue is empty and process log is {sleep_time} seconds.")
         if not process_log:
             Log.warning(f"No work found within {sleep_time} seconds.")
         return process_log

--- a/test/integration/commands/test_autosubmit_commands.py
+++ b/test/integration/commands/test_autosubmit_commands.py
@@ -52,3 +52,28 @@ def test_create_noplot_calls_generate_output(as_exp, mocker, noplot):
         mock_generate_output.assert_not_called()
     else:
         mock_generate_output.assert_called_once()
+
+
+@pytest.mark.parametrize("noclean,uncompress", [
+    (True, True),
+    (True, False),
+    (False, False),
+])
+def test_archive_and_unarchive(as_exp, mocker, noclean, uncompress):
+    as_exp.autosubmit.create(
+        as_exp.expid,
+        noplot=True,
+        hide=True,
+    )
+
+    assert as_exp.autosubmit.archive(as_exp.expid, noclean, uncompress)
+    assert as_exp.autosubmit.unarchive(as_exp.expid, uncompress)
+
+
+def test_archive_noncreated_experiment(as_exp):
+    assert as_exp.autosubmit.archive(as_exp.expid)
+
+
+def test_unarchive_nonarchived_experiment(as_exp):
+    assert not as_exp.autosubmit.unarchive(as_exp.expid)
+

--- a/test/integration/test_log_recovery.py
+++ b/test/integration/test_log_recovery.py
@@ -157,7 +157,7 @@ def test_wait_until_timeout(prepare_test, local, as_conf, mocker, cleanup_event,
     if recovery_queue_full:
         for i in range(max_items):
             local.recovery_queue.put(Job('t000', f'000{i}', Status.COMPLETED, 0))
-    process_log = local.wait_until_timeout(2)
+    process_log = local.wait_mandatory_time(2)
     assert process_log == result
 
 

--- a/test/integration/test_log_recovery.py
+++ b/test/integration/test_log_recovery.py
@@ -323,7 +323,7 @@ def test_refresh_log_recovery_process(local, autosubmit, as_conf, mocker):
 
     local.clean_log_recovery_process()
 
-    assert local.work_event is not None
+    assert local.work_event is None
 
     p = local.work_event
 

--- a/test/unit/platforms/test_locplatform.py
+++ b/test/unit/platforms/test_locplatform.py
@@ -103,6 +103,8 @@ def test_refresh_log_recovery_process(autosubmit, autosubmit_config, mocker):
     as_conf.misc_data["AS_COMMAND"] = 'run'
 
     local = LocalPlatform(expid='t000', name='local', config=as_conf.experiment_data)
+    local.prepare_process()
+    assert local.work_event is not None
 
     spy = mocker.spy(local, 'clean_log_recovery_process')
     spy2 = mocker.spy(local, 'spawn_log_retrieval_process')
@@ -114,4 +116,4 @@ def test_refresh_log_recovery_process(autosubmit, autosubmit_config, mocker):
         autosubmit.refresh_log_recovery_process(platforms=[local], as_conf=as_conf)
         spy.assert_called()
         spy2.assert_not_called()
-        assert local.work_event.is_set()
+        assert local.work_event is None

